### ISSUE TITLE
Tensorflow (TF) Serving on Multi-GPU box 

### ISSUE THREAD

**Issue by Immexxx** [tensorflow/serving#311]
1. What is the best way to serve requests from TF Serving in a multi-GPU box. TF is able to see all the GPUs but uses only one (Default Inception v3 model - from ~ 03/2016 I believe). (Is there some configuration in TF Serving that I am not aware of.)  

Is this dependent on how the model is loaded and exported? 

For inference (and not training) - Is there an example of a saved model being loaded onto multiple GPUs (with a single CPU) where the CPU splits the GPU load among multiple GPUs instead of using only one GPU.

2. If the model is trained on a single GPU machine, when it is being used for inference in TF Serving, can it be used on a multi GPU box where all GPUs are being used?

**Comment by jorgemf** [comment]
TF Serving only executes the graph loaded. If you graph is using multiple GPUs TF serving will use them, if you graph is only using one GPU (most common case) then you will be able to run it in one GPU in TF Serving. Probably your best solution is to build a script which loads your graph once per GPU, then uses some code in the CPU to split the batch data into the GPU graphs and finally export the whole graph with support for multiGPU.

I am not aware of other solution as it could be to replicate the same graph in every GPU in TF serving, but I guess it could be possible by modifying the serving code. Although I don't think this will be supported anytime soon and you will have to do it by yourself.

**Comment by jlertle** [comment]
Using CUDA_VISIBLE_DEVICES environment variable you can pin Model Server processes to a specific GPU. Run each of them on a separate port and throw a load balancer in front.

**Comment by Immexxx** [comment]
@jlertle : Thanks for the comment.  But even with that, if it is a 4 GPU box, you would have to run 4 servers - which does not seem like the most efficient way to do it. Also, you have to run this on 4 different ports, which in turn adds routing complexity.

**Comment by bhack** [comment]
@Immexxx I think that there are not so much resources invested in serving and ecosystem repositories comparing to TF. It is comprensible cause this are a little bit in conflict with selling managed services.

**Comment by Immexxx** [comment]
Guessing I could close this issue based on the above responses, but will be cool for the system to "intelligently" figure out load - for example, a graph requiring only one GPU - and scale it based on available resources - if multiple GPUs are available, replicate the graph - and serve it. 

Could be a flag when launching TF Serving. 

In summary, will leave this open for now - in case someone (including me)  is looking for an interesting project. In general, TF's handling of GPUs - grabbing all the memory of all GPUs and 'locking' them up but using only a subset for compute does not seem like the most elegant way of doing this. A flag for "auto-replicate and serve" would be nice.

**Comment by vitalyli** [comment]
Let's re-open this; With 8GPU machine and VIP in front of a cluster of machines, it would be really helpful if the sever was simply alternating requests across N available GPUs.
For most of us without TPUs, the only way to speed things up is using GPUs. The difference can be 5x faster on GPUs; On a machine with 8 GPUs it's really useful to have server to multiplex requests across all available GPUs.

**Comment by zacharynevin** [comment]
+1. This would be incredibly useful. Currently I have a model that I am trying to run on a `p2.16xlarge` instance but its only utilizing one of the GPUs. It would be incredibly useful if the batch could be split automatically between all the GPUs.

**Comment by zacharynevin** [comment]
In the end I simply used client-side load balancing and had a separate tensorflow serving server for each GPU

**Comment by vitalyli** [comment]
How did you limit instance per GPU. I guess I need to use similar setup.

**Comment by gawinghe** [comment]
me too. tensorflow serving is a wasting of multi gpus

**Comment by mattfeury** [comment]
another +1 here for the exact use case as above. running this on a p2 instance with extra gpus would be an incredibly easy way to scale up. but right now it requires an instance per gpu

**Comment by rundembear** [comment]
The problem is NOT with list_local_devices!! I ran my code once, got the result we use, which is just the string description. Then I commented out the call to list_local_devices and I still get the same crash with the same error:

F tensorflow/core/common_runtime/gpu/gpu_id_manager.cc:45] Check failed: cuda_gpu_id.value() == result.first->second (1 vs. 0)Mapping the same TfGpuId to a different CUDA GPU id. TfGpuId: 0 Existing mapped CUDA GPU id: 0 CUDA GPU id being tried to map to: 1

I am going to try some earlier versions of Tensorflow to see if I can identify where this went bad.

**Comment by t27** [comment]
This issue still appears on the latest versions.

I'm using an AWS g3.8xlarge instance which has 2 GPUs.

TF serving is able to detect both GPUs and initialise them but while running the model it only uses 1 GPU to the maximum.

We are on version 1.7, even though the client sends upto 32 requests in parallel, the model server only uses the first GPU (check screenshot from nvidia-smi)
![06_09_21](https://user-images.githubusercontent.com/2003371/40657873-faca0f68-6366-11e8-963c-3d5ba1db4e2c.jpg)

Seeing that this ticket is open since quite some time, Is the external load balancer solution(that @zacharynevin suggested)the only solution at the moment? 

@rickragv

**Comment by qiaohaijun** [comment]
```bash
Mapping the same TfGpuId to a different CUDA GPU id. TfGpuId: 0 Existing mapped CUDA GPU id: 0 CUDA GPU id being tried to map to: 1
```
when I use multi GPU in tf1.6, I met the error

**Comment by sugartom** [comment]
@jlertle @zacharynevin , hi, I would like to ask what commands you were using to run a separate tensorflow serving server for each GPU.

I have a machine with two 1080 Tis. My TF-Serving is able to correctly identify both of the GPUs when the `CUDA_VISIBLE_DEVICES` flag is not set. And I was trying to run one TF-Serving server per GPU, so I opened two terminals to run the following two commands:

> CUDA_VISIBLE_DEVICES=0 bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server --port=9000 --model_name=inception --model_base_path=/path/to/inception_model

> CUDA_VISIBLE_DEVICES=1 bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server --port=9001 --model_name=mnist --model_base_path=/path/to/mnist_model

The first command was running fine, but the second one will reach error says

> terminate called after throwing an instance of 'std::system_error'
  what():  Resource temporarily unavailable
[1]    4021 abort (core dumped)  CUDA_VISIBLE_DEVICES=1  --port=9001 --model_name=mnist

And idea or suggestions? Thanks in advance!

**Comment by zacharynevin** [comment]
Hi @sugartom, can you post the output of `nvidia-smi`?

**Comment by sugartom** [comment]
@zacharynevin , thanks for your reply, my output of nvidia-smi is:

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.59                 Driver Version: 390.59                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 108...  Off  | 00000000:01:00.0  On |                  N/A |
| 23%   31C    P8    15W / 250W |    334MiB / 11176MiB |     28%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 108...  Off  | 00000000:02:00.0 Off |                  N/A |
| 23%   34C    P8    15W / 250W |      2MiB / 11178MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1089      G   /usr/lib/xorg/Xorg                           225MiB |
|    0      1892      G   compiz                                       103MiB |
|    0      5429      G   /usr/lib/firefox/firefox                       2MiB |
+-----------------------------------------------------------------------------+

**Comment by jlertle** [comment]
I'm currently on v1.6 but have used this method since ~v0.8. 

Have you tried running only on the second GPU? 

Perhaps try this:

export CUDA_VISIBLE_DEVICES=1 && bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server --port=9001 --model_name=mnist --model_base_path=/path/to/mnist_model

**Comment by sugartom** [comment]
Yes, already tried that. My TF-Serving runs perfectly on either gpu:0 or gpu:1 alone, and that's why it confused me so much that running two TF-Serving instances didn't work...

**Comment by Immexxx** [comment]
@sugartom 
1. It is certainly possible to run two instances (at least on a P100 and V100) 
2. Can you try swapping the models to see if there is some issue with the model - like running mnist on GPU 0, to rule out model/model-environment issues 
3. If that does not work, you can try swapping the ports / using different ports  
4. Do you see this error when you are trying to start up the TF Serving or when you are trying to process an image through it? If it is the former and you are still seeing the error, you could try setting the gpu_memory_fraction to see if that helps

**Comment by harshini-gadige** [comment]
Is this still an issue ?

**Comment by Immexxx** [comment]
Yes, I still think it is (but happy to be corrected) 

Currently, TF Serving does not seem to be able to exploit the compute on all on GPU nodes if the graph is set up to run on only one GPU. 
Having said that, it depends on the direction of evolution of TF Serving and if this does not make sense for TF Serving, we can close the issue. 

Given the cost of GPUs and the way GPUs are being made available by vendors and cloud providers, this would be nice to have

**Comment by wydwww** [comment]
@Immexxx @jorgemf 
> If you graph is using multiple GPUs TF serving will use them, if you graph is only using one GPU (most common case) then you will be able to run it in one GPU in TF Serving.

Do you mean when a model is trained in model parallelism, then it can be served with multiple GPUs?

**Comment by wydwww** [comment]
@Immexxx 
Regarding to your latest post,

> Currently, TF Serving does not seem to be able to exploit the compute on all on GPU nodes if the graph is set up to run on only one GPU.

According to Figure 4 of a ICML2017 [paper](http://proceedings.mlr.press/v80/gao18a/gao18a.pdf) about device placement, utilising 2 or 4 GPUs will only sightly reduce latency (from 0.1 to ~0.07) comparing with single-GPU in model parallelism training. 

Is this the goal of utilising multiple GPUs for one request, which is to reduce the minimal latency?

Thanks

**Comment by srikanthcognitifAI** [comment]
In the TensorRT github page they have mentioned "Multi-GPU support. TRTIS can distribute inferencing across all system GPUs." ,is this possible with TFserving already??

**Comment by xuzheyuan624** [comment]
How to use an appointed GPU to run tensorflow serving with docker.I dont't want to take uo all gpus when running tf serving.So does anybody know?

**Comment by aaroey** [comment]
@xuzheyuan624 could you try to set environment variable [CUDA_VISIBLE_DEVICES](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars) and see if it works?

**Comment by xuzheyuan624** [comment]
@aaroey I tried like this:
`#!/bin/bash

export CUDA_VISIBLE_DEVICES=5 && \
docker run --runtime=nvidia -p 8500:8500 \
--mount type=bind,\
source=/home/xzy/keras-yolov3/yolov3,\
target=/models/yolov3 \
-e MODEL_NAME=yolov3 -t tensorflow/serving:1.12.0-gpu &`
but it didn't work.

**Comment by t27** [comment]
You need to use the `nvidia` runtime for docker, and export the `NVIDIA_VISIBLE_DEVICES` environment variable *inside* the container when you run it

Sample command
```
docker run  --runtime=nvidia -e NVIDIA_VISIBLE_DEVICES=0 -p 8501:8501 \
  --mount type=bind,source=/path/to/my_model/,target=/models/my_model \
  -e MODEL_NAME=my_model -t tensorflow/serving:latest-gpu
```

If you use a config file for tensorflow serving, you can use the following command while running the docker 

```
docker run --runtime=nvidia -e NVIDIA_VISIBLE_DEVICES=0 -p 8500:8500 -p 8501:8501 \
  --mount type=bind,source=/path/to/my_model/,target=/models/my_model \
  --mount type=bind,source=/path/to/my/models.config,target=/models/models.config \
  -t tensorflow/serving:latest-gpu --model_config_file=/models/models.config
```

Source: https://github.com/tensorflow/serving/blob/master/tensorflow_serving/g3doc/docker.md#serving-with-docker-using-your-gpu

**Comment by t27** [comment]
Until tensorflow serving incorporates the multi gpu load balancing, the best approach to do multi gpu inference for TensorflowServing is by having dockerised tensorflow serving containers for each gpu (https://github.com/tensorflow/serving/issues/311#issuecomment-480176078) and then writing your own load balancing logic which has a list of all the url-port number combinations and dispatches requests accordingly

**Comment by xuzheyuan624** [comment]
@t27 thanks! It's very useful!

**Comment by troycheng** [comment]
Did someone meet throughput decreasing problem when using load balancing over multiple tf-serving instances?  For only one instance, using batching with fine tuning it can reach 500 qps, but for 8 intances, It can't reach 500 * 8 = 4000 qps, even  worse, it can't reach 500 qps. I know batching is a main cause but till now I do not why and also not find a better solution to make full use of multiple gpu devices

**Comment by 372046933** [comment]
@troycheng 
I met up with the same problem. See https://github.com/kubernetes/kubernetes/issues/87277 and https://github.com/kubernetes/kubernetes/issues/88986. Can you paste your `batching_parameters_file` and the CPU limit on each instance. Pay attention to `num_batch_threads` especially. If it's bigger than the available CPUs. Most of the resources are busy waiting for `condition_variable`. See https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/core/kernels/batching_util/shared_batch_scheduler.h#L497. We have almost fixed the scale up problem by tuning the batching parameter. But the scale up is still not linear(About 0.85* NUM_INSTANCE). I changed the wait time and now I'm recompiling serving to see if there is any improvement.

**Comment by tobegit3hub** [comment]
If we only serve one model with one model version in multiple-gpu server, we have implemented the functionality to automatically start multiple session to load the model which has been edit for binding different gpu devices. This can utilize all the gpu devices for one model and rebalance the inference requests by simple strategies like RoundRobin.

However, the default implementation of servable in TensorFlow Serving manages only one Session by use TensorFlow C API to load SavedModel. We have to implement the new Servable or SourceAdapter to support this. 

There are some issues for multiple models or multiple model versions if we allocate all gpu resource to one model version. It is related to the usage and design of this project. We are willing to contribute what we have done if we have the clear design for multiple gpu support.

**Comment by Uncle-Yuanl** [comment]
> If we only serve one model with one model version in multiple-gpu server, we have implemented the functionality to automatically start multiple session to load the model which has been edit for binding different gpu devices. This can utilize all the gpu devices for one model and rebalance the inference requests by simple strategies like RoundRobin.

Thank you for your information. So can I find a demo or related code in this  project??
Or can you explain the details of the method above.

Thank you very much!

**Comment by Uncle-Yuanl** [comment]
@tobegit3hub 


> If we only serve one model with one model version in multiple-gpu server, we have implemented the functionality to automatically start multiple session to load the model which has been edit for binding different gpu devices. This can utilize all the gpu devices for one model and rebalance the inference requests by simple strategies like RoundRobin.

Thank you for your information. So if there is already a demo or related code in this project??
Or can you explain the details of the method above.

Thank you very much!

**Comment by Antonio-hi** [comment]
This is still an issue

**Comment by raminmohammadi** [comment]
I am facing the same problem on TF 2.11, have to GPUs but its only picking one during the serving.
